### DISCUSSION TITLE
mystified by how `Depends` is implemented

### DISCUSSION THREAD

**Discussion by CharlesJQuarra** [fastapi/fastapi#3232] [Questions] [answered]
I've been trying to understand how `Depends` uses the `get_db` generator under the hood to retrieve the `SessionLocal`, but I must admit defeat and ask for clarification

At first I assumed that it was some special PEP magic sauce that allowed special behavior when generators were returned as default argument values

But this example shows that my assumption was wrong, this fails, because the generator has not been initialized yet with `next`

```

def gen():
  try:
    yield {'foo':[]}
  finally:
    print("Done")
    
def foo(g = gen()):
   print("foo, g = {}".format(g))
```

My second assumption was that `fastapi.Depends` hide some special behavior trick for generators, but looking at the definition at https://github.com/tiangolo/fastapi/blob/master/fastapi/params.py#L352 I am more confused, as this is basically just setting the function as attribute and doing literally nothing with it

**Comment by flashgorman-chorus** [comment]
Try:

```python
def foo(g = Depends(gen)):
    print(...)
```

1. Yes, `fastapi.depends` hides some special behavior.
2. Notice you are passing a reference to your function `gen`, not actually _calling_ `gen`.

When you look at the source code of `Depends`, yes, it is just storing a reference to the function you say you depend on.  At runtime, as part of request dispatching, FastAPI will "see" this dependency you've declared, and call the function you say you depend on, call it on your behalf, and pass whatever value it returns to your function.

**Reply by CharlesJQuarra** [reply]
Please correct me if wrong, but after browsing the code some more, I think here is where the dependency generators are being initialized and evaluated for the 1st time, which is itself being called from `fastapi.routing.get_request_handler`:

https://github.com/tiangolo/fastapi/blob/aece74982d7c9c1acac98e2c872c4cb885677fc7/fastapi/dependencies/utils.py#L544

**Comment by tiangolo** [comment, marked as answer]
Underneath, sort of "in short", FastAPI uses https://docs.python.org/3/library/contextlib.html#contextlib.asynccontextmanager, with some extra tricks for compatibility with Python versions. But it's pretty much that for the case of dependencies with `yield`.

For other dependencies, it just calls the function passing the declared parameters. The value for the parameters would be extracted from the request, just as for path operation functions. And if some of the parameters are dependencies, those dependencies would be called first.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
